### PR TITLE
[wablib] automate -wablib in bnd-maven-plugin

### PR DESCRIPTION
**Pull request by rotty3000**

### CODE REVIEW

**Comment by rotty3000** [comment]
this was too easy not to implement.

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java [hunk 1]
```diff
@@ -226,8 +232,21 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 					}
 					builder.updateModified(cpe.lastModified(), cpe.getPath());
 					buildpath.add(cpe);
+
+					if ((isWab && !hasWablibs) && !Artifact.SCOPE_PROVIDED.equals(artifact.getScope())
+						&& !Artifact.SCOPE_TEST.equals(artifact.getScope()) && !artifact.isOptional()) {
+
+						String fileNameMapping = MappingUtils
+							.evaluateFileNameMapping(MappingUtils.DEFAULT_FILE_NAME_MAPPING, artifact);
+						wablibs.add("WEB-INF/lib/" + fileNameMapping + "=" + cpe.getName() + ";lib:=true");
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: Should `-wablib` be set to just the paths to the jar files? And Bnd will put them in WEB-INF/lib? For example, see https://github.com/bndtools/bnd/blob/61ed372bad81a7564dc1aa0195edd37502e8d601/biz.aQute.bndlib.tests/test/test/BuilderTest.java#L1628-L1650

Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: we can't, the names are different on the classpath than what the artifacts will be in the WEB-INF/lib, we need to use a resolver to do the mapping, just like I did.

Comment (bjhargrave) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: Well, it does not really matter what the names are inside the bundle :-) Just that they are there and on the BCP. But this is not important.

Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: sorry, what I mean to say is that the maven-jar-plugin will assign names to artifacts it places in the WAR, if we try to play with those names, we may end up with two jars of of the same thing in the WAR, which we don't want. So it's best to make sure we preserve these names and do the proper mapping.

Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: for instance, if you have a dependency which is locally built in the same project, artifact.getFile() will return a file who's name doesn't contain the version, BUT maven-war-plugin will shove a jar into the WEB-INF/lib/ that DOES have a version, which means if we also write a jar with no version into WEB-INF/lib then you'll have 

```
WEB-INF/lib/foo.jar
WEB-INF/lib/foo-3.4.6-SNAPSHOT.jar
```

**Comment by rotty3000** [comment]
@bjhargrave updated!

File: maven/bnd-maven-plugin/README.md [hunk 2]
```diff
@@ -167,17 +148,15 @@ The plugin has 3 distinct usage scenarios:
...
+3. The third case is when the project packaging is `war`, for instance when the plugin is paired with the `maven-war-plugin` instead of the `maven-jar-plugin`. The output of the plugin re-directed to the default assembly directory of the `maven-war-plugin`, which is `${project.build.directory}/${project.build.finalName}`:
 
    ```xml
    <plugin>
        <groupId>biz.aQute.bnd</groupId>
        <artifactId>bnd-maven-plugin</artifactId>
        <configuration>
-           <outputDir>${project.build.directory}/${project.build.finalName}</outputDir>
            <bnd><![CDATA[
                Web-ContextPath: /${project.build.finalName}
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/README.md]: It is wrong to show the `<bnd>` example lines having leading spaces. Bnd will interpret this as a line continuation.

File: maven/bnd-maven-plugin/README.md [hunk 3]
```diff
@@ -61,7 +52,7 @@ pom. Bnd instructions in the pom are not used if the project has a bnd file.
  org.example.api,\
  org.example.types
 -sources: true
-]]></bnd>
+        ]]></bnd>
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/README.md]: It is wrong to show the `<bnd>` example lines having leading spaces. Bnd will interpret this as a line continuation.

File: maven/bnd-maven-plugin/README.md [hunk 4]
```diff
@@ -137,7 +116,9 @@ The plugin has 3 distinct usage scenarios:
    </plugin>
    ```
 
-2. The second case, let's call it the _filtering_ use case, is a more advanced usage emulating the behavior found in the Bnd _Workspace_ model where instructions are used to include contents in the resulting jar. Note that controlling the actual contents of the jar is handled through configuration of the `maven-jar-plugin`:
+2. The second case, let's call it the _filtering_ case, is a more advanced usage emulating the behavior found in the Bnd _Workspace_ model where instructions are used to include (i.e. *filter*) contents in the jar.
+
+   **Note** that controlling the actual contents of the jar is handled through configuration of the `maven-jar-plugin`:
 
    ```xml
    <plugin>
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/README.md]: It is wrong to show the `<bnd>` example lines having leading spaces. Bnd will interpret this as a line continuation.

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java [hunk 5]
```diff
@@ -85,15 +88,18 @@
 	@Parameter(defaultValue = "${project.build.resources}", readonly = true)
 	private List<org.apache.maven.model.Resource>	resources;
 
-	@Parameter(defaultValue = "${project.build.outputDirectory}")
+	@Parameter(defaultValue = "${project.build.outputDirectory}", readonly = true)
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: Why are these back to readonly=true? I thought we expect people to be able to configure this?

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java [hunk 6]
```diff
@@ -194,9 +200,24 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 				}
 			}
 
+			boolean isWab = PACKAGING_WAR.equals(project.getPackaging());
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: We are now conflating `-wab` with packaging type = war. What if someone uses `-wab` with packaging type = jar?

File: maven/bnd-maven-plugin/README.md (hunk 2 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/README.md]: it works perfectly in practice!

File: maven/bnd-maven-plugin/README.md (hunk 3 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/README.md]: works perfectly in practice...

File: maven/bnd-maven-plugin/README.md (hunk 4 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/README.md]: works perfectly in practice...

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java (hunk 5 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: .. I thought we just wanted to use the flag to say don't add this to the jar automatically, but I can make it configurable if you like.

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java (hunk 6 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: if someone uses -wab with jar they'll get messed up jar. Should I block the combination with an error?

File: maven/bnd-maven-plugin/README.md (hunk 2 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/README.md]: do you really think this needs to be changed? It's really ugly in the POMs when you do that... meanwhile there is no evidence that it causes any problem.

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java (hunk 6 above)
Comment (rotty3000) [reply maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: my last update solves this concern by throwing an error (like it previously did) when pairing `-wab` or `-wablib` with packaging `jar`.

File: maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java [hunk 7]
```diff
@@ -194,9 +200,28 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+				if (wabProperty == null) {
+					builder.setProperty(Constants.WAB, "");
+				}
+				outputDir = warOutputDir;
+				logger.info(
+					"WAB mode enabled. Bnd output will be expanded into the 'maven-war-plugin' default <webappDirectory>:"
+						+ outputDir);
+			}
+			else if ((wabProperty != null) || hasWablibs) {
+				throw new MojoExecutionException(
```
Comment (bjhargrave) [inline maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java]: I think this should be `MojoFailureException`.

**Comment by rotty3000** [comment]
updated!

**Pull request merged**